### ISSUE TITLE
Changing title text of a block loses variadic "input list:" inputs in callers.

### ISSUE THREAD

**Issue by brianharvey** [jmoenig/Snap#2760]
Normally when you change the title text of a block, its callers are updated to match, without losing the input expressions in the call.  But when the block has a variadic input, and a caller gives it an input list instead of discrete inputs, that input list expression silently disappears in the calling script.

**Comment by jmoenig** [comment]
Okay, I have to take a deep breath so not to get screaming mad. I have resisted adding the "with input list:" gimmick kicking and screaming. You never convinced me but "talked me into" adding this feature which I hate more than anything else, you basically didn't twist my arm but broke it, it's one of the things I don't forget and certainly don't easily forgive. I so hate adding the words "with input list:", it's one of the most ugly, evil, bad UI behaviors ever, and it doesn't help anybody understand anything, and did I mention that I hate it. This is one of the things we will forever disagree on, and it's really ironic: Me, coming from Smalltalk where every method has a really long selector name wishing for short, concise block labels, and you, coming from LISP with basically cryptic abbreviations for everything believe in "whole sentences with relative subclauses" for block labels.

Adding the stupid text breaks the input restoration mechanism. It breaks a few other things as well, but not in ways apparent to the user, and it is detrimental to performance as well, so I won't get into the details, except that I hate everything about it.

There is a hidden setting: You can turn those "dynamic input labels" off, and then the input gets properly restored:

<img width="363" alt="dynamic input labels" src="https://user-images.githubusercontent.com/2596285/103343621-c8d1c680-4a8c-11eb-83a0-be0d8ecace06.png">

This is how Snap! should look and behave wrt input lists.

I'm going to remove the "bug" label and replace it with something else because I can't stand getting a bug report for something I hate so profoundly in the first place.

**Comment by brianharvey** [comment]
Wow.  I started reading your response and was most of the way through it before I understood that it wasn't the input-list feature you hate, but merely the text representing it.  I don't remember having an argument about that, and it's certainly news to me that I "broke your arm."  And I'm pretty sure I've never seen your proposed representation before, although there's a lot I've forgotten.

But, Jens, what other resentments like this are you harboring silently?  As I said before, I feel like I'm always about to step on a landmine with you lately.  (As I'm typing this, I just got another notification about you resenting my using the word "should" in a bug report.)

I'd be happy to talk about a different visual representation of the input list variant to variadic input slots.  But I think your version is way too subtle.  A user reading another user's code is, I think, not going to notice the _absence_ of arrowheads.

Could you narrow down for me what does and doesn't cause problems?  For example, is the problem because the text comes _before_ the input slot?  Could we superimpose a "do not" circle-slash over the arrowheads, instead of just disappearing them?  That would give a user reading someone else's code a hint that _something_ different is going on.

**Comment by brianharvey** [comment]
I mean, jeez, "it's one of the things I don't forget and certainly don't easily forgive."

I've been working on the manual and the libraries, and I ran into a few problems, and I reported them.  I wasn't expecting hate mail back!  Should I stop reporting problems?

**Comment by jmoenig** [comment]
Look, the argument about replacing a variadic input slot with a reporter was a huge argument spanning several months and at the time (8 years ago) was presented to me as a life-or-death decision between using Snap! in schools or not, so I gave in and changed our initial design. I still believe that `join input list: (a list of strings)` is less clear than `join (a list of strings)`, and I've failed to convince you that "input list:" is even the wrong terminology, because it propagates the mental model that it refers to the inputs of the particular *block* instead of - correctly - the elements of a single *input slot* of type "multiple". Because of this naming and substitution scheme it has become unwieldy to make a block with a variadic  input *followed* by one or more other - or worse likewise variadic - inputs, and we're making it seem as if the *block* itself were variadic, which isn't true.

But this has long since passed and now there are a bazillion materials so we're not going to change anything about it. That doesn't make me any happier though, especially since this awful kludge (replacing the variadic input slot with another label element that adds some text and ends up performing the identity function on its single input at evaluation time) with all its necessary follow-up kludges (special-casing the evaluator and the serializer, removing the single input has to remove the label element and somehow restore the variadic input slot again, etc. etc.) adds so much complexity, maintenance cruft and overall error-prone-ness, like this issue here.

TL;DR: Yes, it's a bug.

**Comment by brianharvey** [comment]
If for eight years you have been holding on to considering my eight-years-ago idea about how to represent a list of inputs for a variadic slot as _unforgivable,_ I don't see how I can continue to work on Snap<i>!</i>.  I have no idea what else you've been silently resenting, so I don't see how I can avoid the landmines.

I'm not saying this to try to force you to do anything.  Do what you like.  I'm sorry I hurt you; it was entirely unintentional.  And a big surprise to me.
